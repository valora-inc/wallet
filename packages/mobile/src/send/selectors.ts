import { createSelector } from 'reselect'
import { localCurrencyToUsdSelector } from 'src/localCurrency/selectors'
import { RootState } from 'src/redux/reducers'
import { tokensWithTokenBalanceSelector } from 'src/tokens/selectors'

export const getRecentPayments = (state: RootState) => {
  return state.send.recentPayments
}

export const isSendingSelector = (state: RootState) => {
  return state.send.isSending
}

export const canSendTokensSelector = createSelector(
  tokensWithTokenBalanceSelector,
  localCurrencyToUsdSelector,
  (tokensList, usdExchangeRate) => {
    return tokensList.length > 0 && usdExchangeRate !== null
  }
)
