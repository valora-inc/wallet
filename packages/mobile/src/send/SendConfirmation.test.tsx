import { StackScreenProps } from '@react-navigation/stack'
import BigNumber from 'bignumber.js'
import * as React from 'react'
import { fireEvent, flushMicrotasksQueue, render } from 'react-native-testing-library'
import { Provider } from 'react-redux'
import { ErrorDisplayType } from 'src/alert/reducer'
import { SendOrigin } from 'src/analytics/types'
import { ErrorMessages } from 'src/app/ErrorMessages'
import i18n from 'src/i18n'
import { AddressValidationType, E164NumberToAddressType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import { RootState } from 'src/redux/reducers'
import { getSendFee } from 'src/send/saga'
import SendConfirmation from 'src/send/SendConfirmation'
import { Currency } from 'src/utils/currencies'
import {
  amountFromComponent,
  createMockStore,
  getMockStackScreenProps,
  RecursivePartial,
} from 'test/utils'
import {
  mockAccount2Invite,
  mockAccountInvite,
  mockE164NumberInvite,
  mockInviteTransactionData,
  mockTransactionData,
} from 'test/values'

// A fee of 0.01 cUSD.
const TEST_FEE_INFO_CUSD = {
  fee: new BigNumber(10).pow(16),
  gas: new BigNumber(200000),
  gasPrice: new BigNumber(10).pow(10).times(5),
  currency: Currency.Dollar,
}

// A fee of 0.01 CELO.
const TEST_FEE_INFO_CELO = {
  fee: new BigNumber(10).pow(16),
  gas: new BigNumber(200000),
  gasPrice: new BigNumber(10).pow(10).times(5),
  currency: Currency.Celo,
}

jest.mock('src/components/useShowOrHideAnimation')
jest.mock('src/send/saga')

const mockedGetSendFee = getSendFee as jest.Mock

const mockScreenProps = getMockStackScreenProps(Screens.SendConfirmation, {
  transactionData: mockTransactionData,
  origin: SendOrigin.AppSendFlow,
})

const mockInviteScreenProps = getMockStackScreenProps(Screens.SendConfirmation, {
  transactionData: mockInviteTransactionData,
  origin: SendOrigin.AppSendFlow,
})

type ScreenProps = StackScreenProps<
  StackParamList,
  Screens.SendConfirmation | Screens.SendConfirmationModal
>

describe('SendConfirmation', () => {
  beforeEach(() => {
    mockedGetSendFee.mockClear()
  })

  function renderScreen(
    storeOverrides: RecursivePartial<RootState> = {},
    screenProps?: ScreenProps
  ) {
    const store = createMockStore({
      stableToken: {
        balances: { [Currency.Dollar]: '200', [Currency.Euro]: '100' },
      },
      ...storeOverrides,
    })

    const tree = render(
      <Provider store={store}>
        <SendConfirmation {...(screenProps ? screenProps : mockScreenProps)} />
      </Provider>
    )

    return {
      store,
      ...tree,
    }
  }

  it('renders correctly', async () => {
    const tree = renderScreen()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly for send payment confirmation with cUSD fees', async () => {
    mockedGetSendFee.mockImplementation(async () => TEST_FEE_INFO_CUSD)

    const { getByText, getByTestId } = renderScreen()

    fireEvent.press(getByText('feeEstimate'))

    jest.runAllTimers()
    await flushMicrotasksQueue()

    const feeComponent = getByTestId('feeDrawer/SendConfirmation/totalFee/value')
    expect(amountFromComponent(feeComponent)).toEqual('$0.0133')

    const totalComponent = getByTestId('TotalLineItem/Total/value')
    expect(amountFromComponent(totalComponent)).toEqual('$1.34')
  })

  it('renders correctly for send payment confirmation with CELO fees', async () => {
    mockedGetSendFee.mockImplementation(async () => TEST_FEE_INFO_CELO)

    const { getByText, getByTestId } = renderScreen()

    fireEvent.press(getByText('feeEstimate'))

    jest.runAllTimers()
    await flushMicrotasksQueue()

    const feeComponent = getByTestId('feeDrawer/SendConfirmation/totalFee/value')
    expect(amountFromComponent(feeComponent)).toEqual('0.01')

    // NOTE: CELO fees are currently not combined into the total.
    // TODO: This should equal more than $1.33, depending on the CELO fee value.
    const totalComponent = getByTestId('TotalLineItem/Total/value')
    expect(amountFromComponent(totalComponent)).toEqual('$1.33')
  })

  it('shows a generic `calculateFeeFailed` error when fee estimate fails due to an unknown error', async () => {
    mockedGetSendFee.mockImplementation(async () => {
      throw new Error('Unknown error message')
    })

    const { store, getByText, queryByTestId } = renderScreen()

    store.clearActions()

    jest.runAllTimers()
    await flushMicrotasksQueue()

    const feeComponent = queryByTestId('feeDrawer/SendConfirmation/totalFee/value')
    expect(feeComponent).toBeFalsy()
    expect(getByText('---')).toBeTruthy()

    expect(store.getActions()).toEqual([
      {
        action: null,
        alertType: 'error',
        buttonMessage: null,
        dismissAfter: 5000,
        displayMethod: ErrorDisplayType.BANNER,
        message: i18n.t('calculateFeeFailed', { ns: 'global' }),
        title: null,
        type: 'ALERT/SHOW',
        underlyingError: 'calculateFeeFailed',
      },
    ])
  })

  it('shows an `insufficientBalance` error when fee estimate fails due insufficient user balance', async () => {
    mockedGetSendFee.mockImplementation(async () => {
      throw new Error(ErrorMessages.INSUFFICIENT_BALANCE)
    })

    const { store, getByText, queryByTestId } = renderScreen()

    store.clearActions()

    jest.runAllTimers()
    await flushMicrotasksQueue()

    const feeComponent = queryByTestId('feeDrawer/SendConfirmation/totalFee/value')
    expect(feeComponent).toBeFalsy()
    expect(getByText('---')).toBeTruthy()

    expect(store.getActions()).toEqual([
      {
        action: null,
        alertType: 'error',
        buttonMessage: null,
        dismissAfter: 5000,
        displayMethod: ErrorDisplayType.BANNER,
        message: i18n.t('insufficientBalance', { ns: 'global' }),
        title: null,
        type: 'ALERT/SHOW',
        underlyingError: 'insufficientBalance',
      },
    ])
  })

  it('renders correctly when there are multiple user addresses (should show edit button)', async () => {
    const mockE164NumberToAddress: E164NumberToAddressType = {
      [mockE164NumberInvite]: [mockAccountInvite, mockAccount2Invite],
    }

    const { getByTestId } = renderScreen({
      identity: {
        e164NumberToAddress: mockE164NumberToAddress,
        secureSendPhoneNumberMapping: {
          [mockE164NumberInvite]: {
            addressValidationType: AddressValidationType.FULL,
            address: mockAccount2Invite,
          },
        },
      },
    })

    expect(getByTestId('accountEditButton')).toBeTruthy()
  })

  it('updates the comment/reason', () => {
    const { getByTestId, queryAllByDisplayValue } = renderScreen({
      fees: {
        estimates: {
          send: {
            feeInWei: '1',
          },
        },
      },
    })

    const input = getByTestId('commentInput/send')
    const comment = 'A comment!'
    fireEvent.changeText(input, comment)
    expect(queryAllByDisplayValue(comment)).toHaveLength(1)
  })

  it('navigates to ValidateRecipientIntro when "edit" button is pressed', async () => {
    const mockE164NumberToAddress: E164NumberToAddressType = {
      [mockE164NumberInvite]: [mockAccountInvite, mockAccount2Invite],
    }
    const mockAddressValidationType = AddressValidationType.PARTIAL

    const { getByTestId } = renderScreen({
      identity: {
        e164NumberToAddress: mockE164NumberToAddress,
        secureSendPhoneNumberMapping: {
          [mockE164NumberInvite]: {
            addressValidationType: mockAddressValidationType,
            address: mockAccount2Invite,
          },
        },
      },
    })

    fireEvent.press(getByTestId('accountEditButton'))
    expect(navigate).toHaveBeenCalledWith(Screens.ValidateRecipientIntro, {
      origin: SendOrigin.AppSendFlow,
      transactionData: mockTransactionData,
      addressValidationType: mockAddressValidationType,
    })
  })

  it('does nothing when trying to press "edit" when user has not gone through Secure Send', async () => {
    const mockE164NumberToAddress: E164NumberToAddressType = {
      [mockE164NumberInvite]: [mockAccount2Invite],
    }

    const { queryByTestId } = renderScreen({
      identity: {
        e164NumberToAddress: mockE164NumberToAddress,
        secureSendPhoneNumberMapping: {
          [mockE164NumberInvite]: {
            addressValidationType: AddressValidationType.NONE,
            address: undefined,
          },
        },
      },
    })

    expect(queryByTestId('accountEditButton')).toBeNull()
  })

  it('renders correct modal for invitations', async () => {
    mockedGetSendFee.mockImplementation(async () => TEST_FEE_INFO_CUSD)

    const { queryByTestId, getByTestId } = renderScreen({}, mockInviteScreenProps)

    expect(queryByTestId('InviteAndSendModal')?.props.isVisible).toBe(false)
    fireEvent.press(getByTestId('ConfirmButton'))
    expect(queryByTestId('InviteAndSendModal')?.props.isVisible).toBe(true)
  })
})
