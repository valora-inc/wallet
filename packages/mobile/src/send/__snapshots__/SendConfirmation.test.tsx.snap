// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendConfirmation renders correctly 1`] = `
<RNCSafeAreaView
  style={
    Object {
      "flex": 1,
      "paddingHorizontal": 8,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "height": 56,
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "height": "100%",
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "paddingLeft": 22,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          hitSlop={
            Object {
              "bottom": 10,
              "left": 10,
              "right": 10,
              "top": 10,
            }
          }
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackgroundBorderless",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
        >
          <svg
            height={16}
            style={Object {}}
            testID="BackChevron"
            viewBox="0 0 15 15"
            width={16}
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0.442803 7.08502C0.444561 7.33572 0.533359 7.58609 0.710547 7.78183C0.744524 7.81936 0.780571 7.85359 0.818325 7.88449L6.40331 13.4695C6.79383 13.86 7.41088 13.8447 7.78152 13.4352C8.15216 13.0258 8.13604 12.3773 7.74551 11.9867L2.83181 7.07304L7.74561 2.15923C8.13614 1.76871 8.15225 1.12021 7.78161 0.710762C7.41097 0.301318 6.79393 0.285981 6.4034 0.676505L0.819307 6.2606C0.78116 6.29173 0.744749 6.32625 0.710448 6.36414C0.527591 6.56614 0.438871 6.82633 0.442803 7.08502Z"
              fill="#2E3338"
              style={Object {}}
            />
          </svg>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RCTScrollView
      contentContainerStyle={
        Object {
          "paddingVertical": 10,
        }
      }
      keyboardShouldPersistTaps="handled"
    >
      <View>
        <View
          style={
            Object {
              "marginHorizontal": 16,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "flex-start",
                "paddingBottom": 24,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "flex-start",
                  "flexDirection": "row",
                }
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                    undefined,
                  ]
                }
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "justifyContent": "center",
                      },
                      Object {
                        "backgroundColor": "hsl(0, 53%, 93%)",
                        "borderRadius": 20,
                        "height": 40,
                        "width": 40,
                      },
                    ]
                  }
                >
                  <Text
                    style={
                      Array [
                        Object {
                          "color": "#FFFFFF",
                          "fontFamily": "Inter-Medium",
                          "fontSize": 16,
                        },
                        Object {
                          "color": "hsl(0, 67%, 24%)",
                          "fontSize": 20,
                        },
                      ]
                    }
                  >
                    J
                  </Text>
                </View>
              </View>
              <View
                style={
                  Object {
                    "paddingLeft": 8,
                  }
                }
              >
                <Text
                  style={
                    Object {
                      "color": "#9CA4A9",
                      "fontFamily": "Inter-Regular",
                      "fontSize": 16,
                      "lineHeight": 22,
                    }
                  }
                  testID="HeaderText"
                >
                  sending
                </Text>
                <Text
                  style={
                    Object {
                      "color": "#2E3338",
                      "fontFamily": "Inter-Medium",
                      "fontSize": 16,
                      "lineHeight": 22,
                    }
                  }
                >
                  Jane Doe
                </Text>
              </View>
            </View>
            <Text
              style={
                Array [
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Inter-SemiBold",
                    "fontSize": 32,
                    "lineHeight": 40,
                    "paddingVertical": 8,
                  },
                  Object {
                    "color": "#2E3338",
                  },
                ]
              }
              testID="undefined/value"
            >
              
              $
              1.33
            </Text>
            <TextInput
              allowFontScaling={true}
              autoFocus={false}
              blurOnSubmit={true}
              maxLength={70}
              multiline={true}
              numberOfLines={5}
              onBlur={[Function]}
              onChangeText={[Function]}
              placeholder="addDescription"
              placeholderTextColor="#1AB775"
              rejectResponderTermination={true}
              returnKeyType="done"
              style={
                Object {
                  "alignSelf": "stretch",
                  "color": "#2E3338",
                  "flex": 1,
                  "fontFamily": "Inter-Regular",
                  "fontSize": 18,
                  "height": 100,
                  "lineHeight": 24,
                  "textAlignVertical": "top",
                }
              }
              testID="commentInput/send"
              underlineColorAndroid="transparent"
              value=""
            />
          </View>
          <Modal
            animationType="none"
            hardwareAccelerated={false}
            hideModalContentWhileAnimating={false}
            onModalHide={[Function]}
            onModalWillHide={[Function]}
            onModalWillShow={[Function]}
            onRequestClose={[Function]}
            scrollHorizontal={false}
            scrollOffset={0}
            scrollOffsetMax={0}
            scrollTo={null}
            statusBarTranslucent={true}
            supportedOrientations={
              Array [
                "portrait",
                "landscape",
              ]
            }
            swipeThreshold={100}
            transparent={true}
            visible={false}
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "black",
                  "bottom": 0,
                  "height": 1334,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 750,
                }
              }
            />
            <View
              forwardedRef={[Function]}
              hideModalContentWhileAnimating={false}
              onModalHide={[Function]}
              onModalWillHide={[Function]}
              onModalWillShow={[Function]}
              pointerEvents="box-none"
              scrollHorizontal={false}
              scrollOffset={0}
              scrollOffsetMax={0}
              scrollTo={null}
              statusBarTranslucent={true}
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                  "margin": 37.5,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                }
              }
              supportedOrientations={
                Array [
                  "portrait",
                  "landscape",
                ]
              }
              swipeThreshold={100}
            >
              <RNCSafeAreaView>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "padding": 16,
                      },
                      Object {
                        "borderRadius": 8,
                      },
                      Object {
                        "elevation": 12,
                        "shadowColor": "rgba(156, 164, 169, 0.4)",
                        "shadowOffset": Object {
                          "height": 2,
                          "width": 0,
                        },
                        "shadowOpacity": 1,
                        "shadowRadius": 12,
                      },
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                          "maxHeight": "100%",
                          "padding": 24,
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <RCTScrollView
                    contentContainerStyle={
                      Object {
                        "alignItems": "center",
                      }
                    }
                  >
                    <View>
                      <Image
                        resizeMode="contain"
                        source={
                          Object {
                            "testUri": "../../../packages/mobile/src/images/invite-modal.png",
                          }
                        }
                        style={
                          Object {
                            "height": 100,
                            "marginBottom": 12,
                            "width": 100,
                          }
                        }
                      />
                      <Text
                        style={
                          Object {
                            "color": "#2E3338",
                            "fontFamily": "Jost-Medium",
                            "fontSize": 22,
                            "lineHeight": 28,
                            "marginBottom": 12,
                            "textAlign": "center",
                          }
                        }
                      >
                        inviteAndSendDialog.title
                      </Text>
                      <Text
                        style={
                          Object {
                            "color": "#2E3338",
                            "fontFamily": "Inter-Regular",
                            "fontSize": 16,
                            "lineHeight": 22,
                            "marginBottom": 24,
                            "textAlign": "center",
                          }
                        }
                      >
                        inviteAndSendDialog.body
                      </Text>
                    </View>
                  </RCTScrollView>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "flexWrap": "wrap",
                        "justifyContent": "space-around",
                        "maxWidth": "100%",
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "rippleRadius": undefined,
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      testID="InviteAndSendModal/SecondaryAction"
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#1AB775",
                              "fontFamily": "Inter-SemiBold",
                              "fontSize": 16,
                              "lineHeight": 22,
                            },
                            Object {
                              "color": "#9CA4A9",
                              "paddingTop": 16,
                            },
                          ]
                        }
                      >
                        global:cancel
                      </Text>
                    </View>
                    <View
                      accessible={true}
                      focusable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "rippleRadius": undefined,
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                      testID="InviteAndSendModal/PrimaryAction"
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#1AB775",
                              "fontFamily": "Inter-SemiBold",
                              "fontSize": 16,
                              "lineHeight": 22,
                            },
                            Object {
                              "paddingTop": 16,
                            },
                          ]
                        }
                      >
                        inviteAndSendDialog.button
                      </Text>
                    </View>
                  </View>
                </View>
              </RNCSafeAreaView>
            </View>
          </Modal>
          <Modal
            animationType="none"
            hardwareAccelerated={false}
            hideModalContentWhileAnimating={false}
            onModalHide={[Function]}
            onModalWillHide={[Function]}
            onModalWillShow={[Function]}
            onRequestClose={[Function]}
            scrollHorizontal={false}
            scrollOffset={0}
            scrollOffsetMax={0}
            scrollTo={null}
            statusBarTranslucent={true}
            supportedOrientations={
              Array [
                "portrait",
                "landscape",
              ]
            }
            swipeThreshold={100}
            transparent={true}
            visible={false}
          >
            <View
              accessible={true}
              focusable={true}
              forwardedRef={[Function]}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "black",
                  "bottom": 0,
                  "height": 1334,
                  "left": 0,
                  "opacity": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                  "width": 750,
                }
              }
            />
            <View
              forwardedRef={[Function]}
              hideModalContentWhileAnimating={false}
              onModalHide={[Function]}
              onModalWillHide={[Function]}
              onModalWillShow={[Function]}
              pointerEvents="box-none"
              scrollHorizontal={false}
              scrollOffset={0}
              scrollOffsetMax={0}
              scrollTo={null}
              statusBarTranslucent={true}
              style={
                Object {
                  "flex": 1,
                  "justifyContent": "center",
                  "margin": 37.5,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                }
              }
              supportedOrientations={
                Array [
                  "portrait",
                  "landscape",
                ]
              }
              swipeThreshold={100}
            >
              <RNCSafeAreaView>
                <View
                  style={
                    Array [
                      Object {
                        "backgroundColor": "#FFFFFF",
                        "padding": 16,
                      },
                      Object {
                        "borderRadius": 8,
                      },
                      Object {
                        "elevation": 12,
                        "shadowColor": "rgba(156, 164, 169, 0.4)",
                        "shadowOffset": Object {
                          "height": 2,
                          "width": 0,
                        },
                        "shadowOpacity": 1,
                        "shadowRadius": 12,
                      },
                      Array [
                        Object {
                          "backgroundColor": "#FFFFFF",
                          "maxHeight": "100%",
                          "padding": 24,
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <RCTScrollView
                    contentContainerStyle={
                      Object {
                        "alignItems": "center",
                      }
                    }
                  >
                    <View>
                      <Text
                        style={
                          Object {
                            "color": "#2E3338",
                            "fontFamily": "Jost-Medium",
                            "fontSize": 22,
                            "lineHeight": 28,
                            "marginBottom": 12,
                            "textAlign": "center",
                          }
                        }
                      >
                        encryption.warningModalHeader
                      </Text>
                      <Text
                        style={
                          Object {
                            "color": "#2E3338",
                            "fontFamily": "Inter-Regular",
                            "fontSize": 16,
                            "lineHeight": 22,
                            "marginBottom": 24,
                            "textAlign": "center",
                          }
                        }
                      >
                        encryption.warningModalBody
                      </Text>
                    </View>
                  </RCTScrollView>
                  <View
                    style={
                      Object {
                        "flexDirection": "row",
                        "flexWrap": "wrap",
                        "justifyContent": "space-around",
                        "maxWidth": "100%",
                      }
                    }
                  >
                    <View
                      accessible={true}
                      focusable={true}
                      nativeBackgroundAndroid={
                        Object {
                          "attribute": "selectableItemBackgroundBorderless",
                          "rippleRadius": undefined,
                          "type": "ThemeAttrAndroid",
                        }
                      }
                      onClick={[Function]}
                      onResponderGrant={[Function]}
                      onResponderMove={[Function]}
                      onResponderRelease={[Function]}
                      onResponderTerminate={[Function]}
                      onResponderTerminationRequest={[Function]}
                      onStartShouldSetResponder={[Function]}
                    >
                      <Text
                        style={
                          Array [
                            Object {
                              "color": "#1AB775",
                              "fontFamily": "Inter-SemiBold",
                              "fontSize": 16,
                              "lineHeight": 22,
                            },
                            Object {
                              "paddingTop": 16,
                            },
                          ]
                        }
                      >
                        global:dismiss
                      </Text>
                    </View>
                  </View>
                </View>
              </RNCSafeAreaView>
            </View>
          </Modal>
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "padding": 16,
          "paddingBottom": 8,
        }
      }
    >
      <View>
        <View
          accessible={true}
          focusable={true}
          nativeBackgroundAndroid={
            Object {
              "attribute": "selectableItemBackground",
              "rippleRadius": undefined,
              "type": "ThemeAttrAndroid",
            }
          }
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          testID="feeDrawer/SendConfirmation"
        >
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
              }
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                  undefined,
                ]
              }
            >
              <Text
                style={
                  Object {
                    "color": "#2E3338",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 16,
                    "lineHeight": 22,
                  }
                }
              >
                feeEstimate
              </Text>
              <View
                style={
                  Object {
                    "transform": Array [
                      Object {
                        "rotate": "0deg",
                      },
                    ],
                  }
                }
              >
                <svg
                  fill="none"
                  height={16}
                  viewBox="0 0 16 16"
                  width={16}
                >
                  <path
                    d="M3 6l5 5 5-5"
                    stroke="#B4B9BD"
                  />
                </svg>
              </View>
            </View>
            <View
              style={
                Array [
                  Object {
                    "flexDirection": "row",
                    "justifyContent": "space-between",
                    "marginVertical": 5,
                  },
                  undefined,
                ]
              }
            >
              <View
                style={
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "marginRight": 5,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#2E3338",
                        "fontFamily": "Inter-Regular",
                        "fontSize": 16,
                        "lineHeight": 22,
                      },
                      undefined,
                    ]
                  }
                >
                  
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 5,
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginRight": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
              ]
            }
          >
            total
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="TotalLineItem/Total/value"
          >
            
            $
            1.33
          </Text>
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "marginVertical": 5,
            },
            Object {
              "marginVertical": 0,
            },
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginRight": 5,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                },
              ]
            }
          >
            <View
              accessible={true}
              focusable={false}
              nativeBackgroundAndroid={
                Object {
                  "attribute": "selectableItemBackground",
                  "rippleRadius": undefined,
                  "type": "ThemeAttrAndroid",
                }
              }
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
            >
              <Text
                style={
                  Object {
                    "color": "#9CA4A9",
                    "fontFamily": "Inter-Regular",
                    "fontSize": 14,
                    "lineHeight": 18,
                  }
                }
              >
                <Text
                  style={
                    Array [
                      undefined,
                      Object {
                        "color": undefined,
                      },
                    ]
                  }
                  testID="TotalLineItem/ExchangeRate/value"
                >
                  
                  $
                  0.75
                </Text>
              </Text>
            </View>
          </Text>
        </View>
        <Text
          style={
            Array [
              Object {
                "color": "#2E3338",
                "fontFamily": "Inter-Regular",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "color": "#9CA4A9",
                "fontFamily": "Inter-Regular",
                "fontSize": 14,
                "lineHeight": 18,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="TotalLineItem/Subtotal/value"
          >
            
            $
            1.00
          </Text>
        </Text>
      </View>
    </View>
    <View
      style={
        Object {
          "marginHorizontal": 16,
          "paddingVertical": 16,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "flexDirection": "column",
            },
            Object {
              "marginBottom": 16,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 100,
              },
            ]
          }
        >
          <View
            accessible={true}
            focusable={true}
            nativeBackgroundAndroid={
              Object {
                "attribute": "selectableItemBackground",
                "rippleRadius": undefined,
                "type": "ThemeAttrAndroid",
              }
            }
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#1AB775",
                "flexGrow": 1,
                "height": 48,
                "justifyContent": "center",
                "opacity": undefined,
                "paddingHorizontal": 24,
                "paddingVertical": 5,
              }
            }
            testID="ConfirmButton"
          >
            <Text
              accessibilityLabel="global:send"
              style={
                Object {
                  "color": "#FFFFFF",
                  "fontFamily": "Inter-SemiBold",
                  "fontSize": 16,
                  "lineHeight": 22,
                }
              }
            >
              global:send
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      collapsable={false}
      style={
        Array [
          Object {
            "bottom": 0,
            "left": 0,
            "right": 0,
          },
          Object {
            "height": 0,
          },
          undefined,
        ]
      }
    />
  </View>
</RNCSafeAreaView>
`;
